components/TransactionList/index.js

```diff
@@ -1,13 +1,73 @@
+import { Fragment } from "react";
 import TransactionCard from "../TransactionCard";
-import { Ul } from "./TransactionList.styled";
+import * as Styled from "./TransactionList.styled";
+import {
+  findYearsInTransactions,
+  filterTransactionsByYear,
+  filterTransactionsByMonth,
+  sortMonths,
+} from "../../utils/transaction";
+import { months } from "@/utils/months";
 
 function TransactionList({ transactions }) {
+  const yearsInTransactions = findYearsInTransactions(transactions);
+
+  const transactionsPerYearsAndMonths = yearsInTransactions.map(
+    (year, index) => {
+      const transactionsPerYear = filterTransactionsByYear(transactions, year);
+      if (transactionsPerYear.length === 0) {
+        return null;
+      }
+
+      const monthsToMap =
+        index === 0
+          ? sortMonths(new Date().getMonth())
+          : months.slice().reverse();
+
+      const transactionsPerMonths = monthsToMap.map((month) => {
+        const monthIndex = months.indexOf(month);
+        const transactionsPerMonth = filterTransactionsByMonth(
+          transactionsPerYear,
+          monthIndex
+        );
+        if (transactionsPerMonth.length === 0) {
+          return null;
+        }
+        transactionsPerMonth.sort(
+          (a, b) => new Date(b.date) - new Date(a.date)
+        );
+        return { month, transactions: transactionsPerMonth };
+      });
+
+      return { year, transactionsPerMonths };
+    }
+  );
+
   return (
-    <Ul>
-      {transactions.map((transaction) => (
-        <TransactionCard key={transaction._id} transaction={transaction} />
+    <>
+      {transactionsPerYearsAndMonths.map((year, index) => (
+        <Fragment key={index}>
```

Comment:
please never use the index as a key, this is bad practice. use a unique never changing value as the key. The index might change in the future.